src/tasks/python/PythonExtensionHelper.ts

```diff
@@ -0,0 +1,116 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See LICENSE.md in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+// This will eventually be replaced by an API in the Python extension. See https://github.com/microsoft/vscode-python/issues/7282
+
+import * as fse from 'fs-extra';
+import * as os from 'os';
+import * as path from 'path';
+import * as vscode from "vscode";
+import CliDockerClient from '../../debugging/coreclr/CliDockerClient';
+import { delay } from '../../utils/delay';
+import { PythonDefaultDebugPort, PythonFileTarget, PythonModuleTarget } from '../../utils/pythonUtils';
+
+export namespace PythonExtensionHelper {
+    export interface DebugLaunchOptions {
+        host?: string;
+        port?: number;
+        wait?: boolean;
+    }
+
+    export function getDebuggerEnvironmentVars(): { [key: string]: string } {
+        return { ["PTVSD_LOG_DIR"]: `/dbglogs` };
+    }
+
+    export function getDebuggerLogFilePath(folderName: string) : string {
+    // The debugger generates the log file with the name in this format: ptvsd-{pid}.log,
+    // So given that we run the debugger as the entry point, then the PID is guaranteed to be 1.
+        return path.join(os.tmpdir(), folderName, "ptvsd-1.log");
+    }
+
+    export async function ensureDebuggerReady(prelaunchTask: vscode.Task, debuggerSemaphorePath: string, containerName: string, cliDockerClient: CliDockerClient): Promise<void> {
+        // tslint:disable-next-line:promise-must-complete
+        return new Promise((resolve, reject) => {
+            vscode.tasks.onDidEndTask(async e => {
+                if (e.execution.task === prelaunchTask) {
+                    // There is no way to know the result of the completed task, so a best guess is to check if the container is running.
+                    const containerRunning = await cliDockerClient.inspectObject(containerName, { format: "{{.State.Running}}" });
+
+                    if (containerRunning === "false") {
+                        reject("Failed to attach the debugger, please see the terminal output for more details.");
+                    }
+
+                    const maxRetriesCount = 20;
+                    let retries = 0;
+                    let created = false;
+
+                    // Look for the magic string below in the log file with a retry every 0.5 second for a maximum of 10 seconds.
+                    // TODO: Should be gone as soon as the retry logic is part of the Python debugger/extension.
+                    while (++retries < maxRetriesCount && !created) {
+                        if (fse.existsSync(debuggerSemaphorePath)) {
```

Comment:
Any reason for the synchronous functions over the asynchronous ones?

Reply:
The only reason is that the asynchronous version accepts a callback as a parameter and returns void. So I couldn't find a way to "wait" for it's completion in a non-messy way, thoughts?

Reply:
`fs-extra` has overloads for most functions; if you pass a callback they return `void` and act as a traditional Node asynchronous functions.  But, if you *don't* pass a callback, they return a `Promise` which can then be awaited.

Reply:
I think it's `fse.pathExists(...)` rather than `fse.exists(...)`